File: wick/step_4b_reclassify.py
"""
Step 4b: Reclassify fluctuation-potential one-body pieces.

pdaggerq introduces an intermediate occupied-repulsion object from the j1
piece of the fluctuation potential, then reclassifies it into an ERI by
introducing an extra occupied summation index:

    <?i||?j>  ->  <i,k||j,k>

This step mirrors that behavior for terms carrying tensors with
name == 'occ_repulsion'.
"""

from __future__ import annotations
from typing import List
from .term import Term, Tensor


_OCC_PREF = ["i", "j", "k", "l", "m", "n"]


def _all_labels(term: Term) -> set[str]:
    labels = set()
    for t in term.tensors:
        labels.update(t.indices)
    for d in term.deltas:
        labels.add(d.i)
        labels.add(d.j)
    for op in term.operators:
        labels.add(op.label)
    return labels


def _fresh_occ_label(used: set[str]) -> str:
    for lbl in _OCC_PREF:
        if lbl not in used:
            return lbl
    n = 0
    while True:
        lbl = f"o{n}"
        if lbl not in used:
            return lbl
        n += 1


def reclassify_occ_repulsion(terms: List[Term]) -> List[Term]:
    """Convert occ_repulsion tensors into equivalent ERI tensors."""
    out = []
    for term in terms:
        used = _all_labels(term)
        new_tensors = []
        for t in term.tensors:
            if t.name != "occ_repulsion":
                new_tensors.append(t.copy())
                continue
            i, j = t.indices
            k = _fresh_occ_label(used)
            used.add(k)
            new_tensors.append(Tensor("g", [i, k, j, k]))
        out.append(Term(term.factor, term.sign, new_tensors, [op.copy() for op in term.operators], [d.copy() for d in term.deltas]))
    return out

File: src/step_4b_reclassify.rs
//! Step 4b: Reclassify fluctuation-potential one-body pieces.
//!
//! Convert occ_repulsion tensors into equivalent ERI tensors:
//!   <?i||?j>  ->  <i,k||j,k>

use std::collections::HashSet;

use crate::term::{Tensor, Term};

const OCC_PREF: &[&str] = &["i", "j", "k", "l", "m", "n"];

fn all_labels(term: &Term) -> HashSet<String> {
    let mut labels = HashSet::new();
    for t in &term.tensors {
        for idx in &t.indices {
            labels.insert(idx.clone());
        }
    }
    for d in &term.deltas {
        labels.insert(d.i.clone());
        labels.insert(d.j.clone());
    }
    for op in &term.operators {
        labels.insert(op.label.clone());
    }
    labels
}

fn fresh_occ_label(used: &HashSet<String>) -> String {
    for lbl in OCC_PREF {
        if !used.contains(*lbl) {
            return lbl.to_string();
        }
    }
    let mut n = 0;
    loop {
        let lbl = format!("o{}", n);
        if !used.contains(&lbl) {
            return lbl;
        }
        n += 1;
    }
}

pub fn reclassify_occ_repulsion(terms: &[Term]) -> Vec<Term> {
    let mut out = Vec::new();
    for term in terms {
        let mut used = all_labels(term);
        let mut new_tensors = Vec::new();
        for t in &term.tensors {
            if t.name != "occ_repulsion" {
                new_tensors.push(t.clone());
                continue;
            }
            let i = &t.indices[0];
            let j = &t.indices[1];
            let k = fresh_occ_label(&used);
            used.insert(k.clone());
            new_tensors.push(Tensor::from_strings(
                "g".to_string(),
                vec![i.clone(), k.clone(), j.clone(), k],
            ));
        }
        out.push(Term::new(
            term.factor,
            term.sign,
            new_tensors,
            term.operators.clone(),
            term.deltas.clone(),
        ));
    }
    out
}
